From c22d62af7e2c5c2a5aaff8dccf45904cf90b9350 Mon Sep 17 00:00:00 2001
From: Sushanth Sowmyan <khorgath@gmail.com>
Date: Wed, 17 Jun 2015 16:34:07 -0700
Subject: [PATCH 0490/1164] CDH-36980 : HIVE-11023 : Disable directSQL if
 datanucleus.identifierFactory = datanucleus2
 (Sushanth Sowmyan, reviewed by Ashutosh Chauhan)

(cherry picked from commit 524cd7948d8c2b16a057e52608deeaa5fb9fa4ac)

Conflicts:
	metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java

Change-Id: I3f7e5077a0d410a92d65cb104de5016c1e377a8c
---
 .../hadoop/hive/metastore/MetaStoreDirectSql.java  |   13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java b/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java
index b5bfcf1..485fc08 100644
--- a/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java
+++ b/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java
@@ -127,9 +127,16 @@ public MetaStoreDirectSql(PersistenceManager pm, Configuration conf) {
         HiveConf.getBoolVar(conf, ConfVars.METASTORE_ORM_RETRIEVE_MAPNULLS_AS_EMPTY_STRINGS);
     defaultPartName = HiveConf.getVar(conf, ConfVars.DEFAULTPARTITIONNAME);
 
-    this.isCompatibleDatastore = ensureDbInit() && runTestQuery();
-    if (isCompatibleDatastore) {
-      LOG.info("Using direct SQL, underlying DB is " + dbType);
+    String jdoIdFactory = HiveConf.getVar(conf, ConfVars.METASTORE_IDENTIFIER_FACTORY);
+    if (! ("datanucleus1".equalsIgnoreCase(jdoIdFactory))){
+      LOG.warn("Underlying metastore does not use 'datanuclues1' for its ORM naming scheme."
+          + " Disabling directSQL as it uses hand-hardcoded SQL with that assumption.");
+      isCompatibleDatastore = false;
+    } else {
+      isCompatibleDatastore = ensureDbInit() && runTestQuery();
+      if (isCompatibleDatastore) {
+        LOG.info("Using direct SQL, underlying DB is " + dbType);
+      }
     }
   }
 
-- 
1.7.9.5

